library/bigip_software.py

```diff
@@ -212,673 +290,1060 @@
-    def factory(module):
-        type = module.params.get('connection')
+        if remote:
+            software = self.handle_remote_source(link)
+            self._values['software_name'] = software
+            if link_md5:
+                image_md5 = self.handle_remote_source(link_md5)
+                self._values['imgmd5_name'] = image_md5
 
-        if type == "rest":
-            module.fail_json(msg='The REST connection is currently not supported')
-        elif type == "soap":
-            if not BIGSUDS_AVAILABLE:
-                raise Exception("The python bigsuds module is required")
-            return BigIpSoapApi(check_mode=module.check_mode, **module.params)
+        else:
+            software = os.path.split(link)[1]
+            self._values['software_name'] = software
 
-    factory = staticmethod(factory)
+        return self._values['software']
 
+    @property
+    def hotfix(self):
+        link = self._values['hotfix']
+        link_md5 = self._values['hotfix_md5sum']
+        remote = self.remote_src
 
-class BigIpCommon(object):
-    def __init__(self, *args, **kwargs):
-        self.result = dict(changed=False, changes=dict())
+        if link is None:
+            return None
 
-        self.current = dict()
+        if remote:
+            hotfix = self.handle_remote_source(link)
+            self._values['hotfix_name'] = hotfix
+            if link_md5:
+                hotfix_md5 = self.handle_remote_source(link_md5)
+                self._values['hfmd5_name'] = hotfix_md5
 
-        if kwargs['hotfix']:
-            kwargs['photfix'] = self.iso_info(kwargs['hotfix'])
+        else:
+            hotfix = os.path.split(link)[1]
+            self._values['hotfix_name'] = hotfix
 
-        if kwargs['software']:
-            kwargs['psoftware'] = self.iso_info(kwargs['software'])
+        return self._values['hotfix']
 
-        self.params = kwargs
+    @property
+    def options(self):
+        state = self._values['state']
+        create = self._values['create_volume']
+        opt = list()
 
-    def _find_iso_content(self, iso):
-        paths = ['/METADATA.XML', 'metadata.xml']
+        if create:
+            opt.append({'create-volume': True})
+        if state == 'activated':
+            opt.append({'reboot': True})
 
-        for path in paths:
-            try:
-                content = iso.record(path).content
-                return content
-            except KeyError:
-                pass
-        raise Exception(
-            "Unable to find metadata file in ISO. Please file a bug"
-        )
+        return opt
 
-    def iso_info(self, iso):
-        result = dict(
-            product=None,
-            version=None,
-            build=None
-        )
+    @property
+    def volume(self):
+        state = self._values['state']
+        volume = self._values['volume']
+        reuse = self._values['reuse_inactive_volume']
+
+        if state == 'activated' or state == 'installed':
```

Comment:
have done the change but this leads to problems like this:

````````
Error
Traceback (most recent call last):
  File "/usr/lib/python2.7/unittest/case.py", line 329, in run
    testMethod()
  File "/usr/local/lib/python2.7/dist-packages/mock/mock.py", line 1305, in patched
    return func(*args, **keywargs)
  File "/home/vagrant/ansible_test/test/unit/bigip/test_bigip_software.py", line 271, in test_activate_installed_volume
    results = mm.exec_module()
  File "/home/vagrant/ansible_test/library/bigip_software.py", line 805, in exec_module
    changed = self.activated()
  File "/home/vagrant/ansible_test/library/bigip_software.py", line 845, in activated
    return self.activate()
  File "/home/vagrant/ansible_test/library/bigip_software.py", line 970, in activate
    self._update_changed_options()
  File "/home/vagrant/ansible_test/library/bigip_software.py", line 833, in _update_changed_options
    attr2 = getattr(self.have, key)
  File "/home/vagrant/ansible_test/library/bigip_software.py", line 624, in volume
    raise F5ModuleError('You must specify a volume.')
F5ModuleError: You must specify a volume.

````````

This is what I have changed it to:

````````
    @property
    def volume(self):
        state = self._values['state']
        volume = self._values['volume']
        reuse = self._values['reuse_inactive_volume']

        if state in ['absent', 'present']:
            return self._values['volume']

        if volume is not None:
            if not self._volume_exists_on_device(volume):
                self._values['create_volume'] = True
            else:
                self._check_active_volume(volume)

            return self._values['volume']

        if volume is None:
            if reuse:
                new_volume = self._get_next_available_volume()
                if self._volume_exists_on_device(volume):
                    self._values['delete_volume'] = True
                    self._values['create_volume'] = True
                else:
                    self._values['create_volume'] = True
                self._values['volume'] = new_volume

                return self._values['volume']
            else:
                raise F5ModuleError('You must specify a volume.')

````````

I managed to overcome this bit by specifying api map for: {name: volume}, this is cause the returned volume json, uses 'name' parameter as volume ID.

However this leads to another issue, now i get this stacktrace on the same test:

``````

Error
Traceback (most recent call last):
  File "/usr/lib/python2.7/unittest/case.py", line 329, in run
    testMethod()
  File "/usr/local/lib/python2.7/dist-packages/mock/mock.py", line 1305, in patched
    return func(*args, **keywargs)
  File "/home/vagrant/ansible_test/test/unit/bigip/test_bigip_software.py", line 271, in test_activate_installed_volume
    results = mm.exec_module()
  File "/home/vagrant/ansible_test/library/bigip_software.py", line 815, in exec_module
    changes = self.changes.to_return()
  File "/home/vagrant/ansible_test/library/bigip_software.py", line 751, in to_return
    result[returnable] = getattr(self, returnable)
  File "/home/vagrant/ansible_test/library/bigip_software.py", line 624, in volume
    raise F5ModuleError('You must specify a volume.')
F5ModuleError: You must specify a volume.

``````
We may have discovered a bug with this change, but I need your help to fix this